src/test/java/com/efficio/etl/web/ImportObservationsControllerTest.java

```diff
@@ -110,9 +113,65 @@ public void testProcessImportWithErrors() throws IOException, WorkbookParserExce
+			.thenReturn(importData);
+		Mockito.when(this.etlService.retrieveCurrentWorkbook(this.userSelection)).thenReturn(workbook);
+		Mockito.when(this.etlService.convertMessageList(ArgumentMatchers.<List<Message>>any())).thenReturn(Arrays.asList("error"));
+		final Map<String, List<Message>> projectDataErrors = new HashMap<>();
+		projectDataErrors.put("ERRORS", new ArrayList<Message>());
+		Mockito.when(this.etlService.validateProjectData(importData, PROGRAM_UUID)).thenReturn(projectDataErrors);
+		Mockito.when(this.etlService.isWorkbookHasObservationRecords(ArgumentMatchers.eq(this.userSelection), ArgumentMatchers.<String>anyList(), ArgumentMatchers.eq(workbook))).thenReturn(true);
+		Mockito.when(this.etlService.isObservationOverMaximumLimit(ArgumentMatchers.eq(this.userSelection), ArgumentMatchers.<String>anyList(), ArgumentMatchers.eq(workbook))).thenReturn(false);
+
+		final String returnValue =
+			this.importObservationsController.processImport(this.uploadForm, 1, this.model, this.session, this.request);
+		Assert.assertEquals("etl/validateProjectData", returnValue);
+		Mockito.verify(this.contextUtil).getCurrentProgramUUID();
+		Mockito.verify(this.etlService).createWorkbookFromUserSelection(ArgumentMatchers.eq(this.userSelection), ArgumentMatchers.anyBoolean());
+		Mockito.verify(this.dataImportService).removeLocationNameVariableIfExists(importData);
+		Mockito.verify(this.dataImportService).assignLocationIdVariableToEnvironmentDetailSection(importData);
+		Mockito.verify(this.dataImportService).checkForInvalidGids(ArgumentMatchers.eq(importData), ArgumentMatchers.<Message>anyList());
+		Mockito.verify(this.etlService).extractExcelFileData(workbook, this.userSelection, importData, true);
+		Mockito.verify(this.etlService, Mockito.times(2)).convertMessageList(ArgumentMatchers.<List<Message>>any());
+	}
+
+	@Test
+	public void testProcessImportWithMisMatchErrors() throws IOException {
+		final Workbook workbook = Mockito.mock(Workbook.class);
+		final org.generationcp.middleware.domain.etl.Workbook importData =
+			Mockito.mock(org.generationcp.middleware.domain.etl.Workbook.class);
+		Mockito.when(this.etlService.createWorkbookFromUserSelection(ArgumentMatchers.eq(this.userSelection), ArgumentMatchers.anyBoolean()))
+			.thenReturn(importData);
+		Mockito.when(this.etlService.retrieveCurrentWorkbook(this.userSelection)).thenReturn(workbook);
+		Mockito.when(this.etlService.convertMessageList(ArgumentMatchers.<List<Message>>any())).thenReturn(Arrays.asList("error"));
+		Mockito.when(this.etlService.headersContainsObsUnitId(importData)).thenReturn(false);
+		final List<String> headers = new ArrayList<>();
+		Mockito.when(this.etlService.retrieveColumnHeaders(workbook, this.userSelection, false)).thenReturn(headers);
+		final Map<String, List<Message>> mismatchErrors = new HashMap<>();
+		mismatchErrors.put("ERRORS", new ArrayList<Message>());
```

Comment:
Same here, can we put checking if these errors were set properly as model attribute?